docs/design/sum_types.md

```diff
@@ -0,0 +1,231 @@
+function, and a value of an empty alternative like `None` is specified by naming
+it:
+
+```carbon
+var my_opt: Optional(i32) = Optional(i32).None;
+my_opt = Optional(i32).Some(42);
+```
+
+The value of a choice type can be inspected using a `match` statement:
+
+```carbon
+match (my_opt) {
+  case .Some(the_value: i32) => {
+    Print(the_value);
+  }
+  case .None => {
+    Print("None");
+  }
+}
+```
+
+## User-defined sum types
+
+`choice` declarations are a convenience shorthand for common use cases, but they
+have limited flexibility. There is no way to control the representation of a
+`choice` type, or define methods or other members for it (although you can
+extend it to implement interfaces, using an
+[`external impl`](generics/overview.md#implementing-interfaces) or
+[`adapter`](generics/overview.md#adapting-types)). However, a `class` type can
+be extended to behave like a sum type. This is much more verbose than a `choice`
+declaration, but gives the author full control over the representation and class
+members.
+
+The ability to create instances of the sum type can be straightforwardly
+emulated with factory functions and static constants, and the internal storage
+layout will presumably involve untagged unions or some other low-level storage
+primitive which hasn't been designed yet, but the key to defining a sum type's
+interface is enabling it to support pattern matching. To do that, the sum type
+has to specify two things:
+
+-   The set of all possible alternatives, including their names and parameter
+    types, so that the compiler can typecheck the `match` body, identify any
+    unreachable `case`s, and determine whether any `case`s are missing.
+-   The algorithm that, given a value of the sum type, determines which
+    alternative is present, and specifies the values of its parameters.
+
+It does so by implementing the `Match` interface, which is defined as follows:
+
+```carbon
+interface Match {
+  let template Continuation:! Type;
+  fn Op[me: Self, C:! Continuation](continuation: Continuation*)
+    -> Continuation.ReturnType;
+}
+```
+
+`Continuation` must itself be an interface, and its definition specifies the set
+of possible alternatives: each alternative is represented as a method of that
+interface. When compiling a proper pattern (or set of patterns that includes a
+proper pattern, as with the cases of a `match`) whose type is a sum type, the
+compiler generates an implementation of `Continuation` and passes it to
+`Match.Op`. The sum type's implementation of `Match.Op` is responsible for
+determining which alternative is present and what its parameters are, and
+calling the corresponding method of `continuation` with those parameters. It is
+required to call exactly one such method exactly once before returning. The
+compiler populates the `Continuation` method bodies with whatever code should be
+executed when the corresponding alternatives match.
```

Comment:
Is there any guarantee as to whether the continuation includes the `match` body or not? Eg, can an implementation choose to call `Match.Op` with a continuation that just returns an index of which match block it should use, and then switch on that after `Match.Op` returns? What guarantees exist about the lifetimes of any local variables in the `Match.Op` function as compared to the `case` block?

There are some optimizations that I'd like to be allowed to perform here (eg, combining multiple matches on different things into a single jump table lookup), but their validity would depend a lot on the rules selected.

Reply:
Hm. I think that with sufficiently clever code generation, we don't need this to be an observable part of the semantics after all. If we say:

-   Matching is performed left-to-right within a pattern, and `match` patterns are considered top-down.
-   Any case block is executed within the call to the continuation in the execution of a `Match.Op` call.
-   If there would be more than one `Match.Op` call for the same scrutinee, the compiler is allowed to omit all but the first.

... then I think I can still do the optimizations that I want to do. (The trick that I have in mind is that the first call to `Match.Op` either runs the code for that alternative, or returns a value saying which alternative was chosen. Subsequent matches on the same thing that can't be merged into that same call can instead look at that returned value, and rerun the `Match.Op` call to re-enter the scope even though they know which alternative will be invoked.)

Reply:
> Is there any guarantee as to whether the continuation includes the `match` body or not?

No, there's no guarantee; I want to err on the side of permitting those kinds of optimizations. I've added a paragraph making that explicit. There are some potential future complications here, though, so I've added a discussion to "Alternatives considered".

I've also added a note that the continuation should be invoked as a tail call, which kind of goes hand in hand with this lack of guarantee.

> What guarantees exist about the lifetimes of any local variables in the `Match.Op` function as compared to the `case` block?

We probably don't want them to be interleaved (i.e. neither disjoint nor nested), if only for debuggability, but I'm not sure we need to specify that. Other than that, there are no guarantees about how the lifetimes relate other than the ones obviously imposed by causality (e.g. the lifetimes of `Match.Op` locals must start before the lifetimes of `case` body locals start).

> * Matching is performed left-to-right within a pattern, and `match` patterns are considered top-down.
> * Any case block is executed within the call to the continuation in the execution of a `Match.Op` call.
> * If there would be more than one `Match.Op` call for the same scrutinee, the compiler is allowed to omit all but the first.

Those constraints all make sense to me (and I think they permit the in-place mutation case described above), but I tentatively think they belong more in the docs for pattern matching than the docs for sum types.